### PR TITLE
gmp 6.3.0: add support for win64

### PR DESCRIPTION
**Pull request by skupr-anaconda**
**Destination channel:** main

### Links

- [PKG-3899](https://anaconda.atlassian.net/browse/PKG-3899) 
- [Upstream repository](https://gmplib.org/repo/gmp-6.3/file/tip)

### Explanation of changes:

- Enable Windows support: add a new `bld.bat` to delegate the build process  to build.sh
- Rewrite `build.sh` to enable `win64` support:
  - the main point is to move files from `"${PREFIX}"/Library/mingw-w64` to `"${PREFIX}"/Library/`
- Remove an outdated [recipe/0000-Avoid-the-x18-register.patch](https://github.com/AnacondaRecipes/gmp-feedstock/pull/3/files#diff-a9788adec7f818df36a133c475ae2f6949aba595f83a822aafd027417e428870)
- Reset the `build/number` to `0`
- Add **posix** and **mingw64** packages and `{{ compiler('m2w64_c') }}`  & `{{ compiler('m2w64_cxx') }}` to `build` on `win`
- Add `--no-error-overlinking` to `abs.yaml`. We got an **error** on our CI but it's a **warning** on the WinDevBoxes: `WARNING (gmp,Library/bin/libgmpxx-4.dll): .. but ['defaults/win-64::m2w64-gcc-libs-core==5.3.0=7'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)`. Seems `gmp` doesn't need `m2w64-gcc-libs-core  # [win]` at runtime
- Add `run_test.bat` to compile the `test.c` file
- Add `license_family` 
- Update `license_file`
- Add `description`

### Notes:

- `r-base 4.3.1` (with the MSYS2/UCRT64 toolchain) requires `gmp` on `win64`. We try to build `r-base` using `gmp` from the defaults channel to make the maintenance easier. 
  - I was able to build `r-base 4.3.1` with this `gmp 6.3.0` locally on the WinDevBoxes and then to build `r-gmp`, `r-rcdd`, `r-cppalgos` and `r-clusterr` (which depends on `gmp` or `gmp`/`r-gmp`). Previously, we could build all of them only with `ucrt64-gmp`.
- To be noticed: The host/test environment can have 3-4 different _gmps_ (from local, m2, m2w64, ucrt64). 

[PKG-3899]: https://anaconda.atlassian.net/browse/PKG-3899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by skupr-anaconda** [comment]
@ifitchet I've tried to build `r-symengine` with `gmp` from the local channel on the WinDevBoxes and it failed:
```
-- Found MPFR: C:/Users/dev-admin/conda-bld/r-symengine_1706623000991/_h_env/Library/lib/mpfr.lib
CMake Error at C:/Users/dev-admin/conda-bld/r-symengine_1706623000991/_h_env/Library/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find GMP (missing: GMP_LIBRARIES)
Call Stack (most recent call first):
  C:/Users/dev-admin/conda-bld/r-symengine_1706623000991/_h_env/Library/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindGMP.cmake:19 (find_package_handle_standard_args)
  CMakeLists.txt:373 (find_package)
```

File: recipe/build.sh [hunk 1]
```diff
@@ -1,48 +1,68 @@
...
-shopt -s extglob
-chmod +x configure
+IFS=$' \t\n' # workaround for conda 4.2.13+toolchain bug
 
-mkdir build
-cd build
+# Adopt a Unix-friendly path if we're on Windows (see bld.bat).
+[ -n "$PATH_OVERRIDE" ] && export PATH="$PATH_OVERRIDE"
 
-if [[ "$target_platform" == "linux-ppc64le" ]]; then
```
Comment (JeanChristopheMorinPerso) [inline recipe/build.sh:12]: We don't build for ppc64 anymore, but I would leave this here for the services team.

File: recipe/build.sh [hunk 2]
```diff
@@ -1,48 +1,68 @@
...
+
+set -e
+
 # Get an updated config.sub and config.guess
 cp $BUILD_PREFIX/share/gnuconfig/config.guess config.fsf.guess
 cp $BUILD_PREFIX/share/gnuconfig/config.sub config.fsf.sub
 
-shopt -s extglob
-chmod +x configure
+IFS=$' \t\n' # workaround for conda 4.2.13+toolchain bug
```
Comment (JeanChristopheMorinPerso) [inline recipe/build.sh]: What's the bug in question?

File: recipe/bld.bat [hunk 3]
```diff
@@ -0,0 +1,18 @@
+:: Trailing semicolon in this variable as set by current (2017/01)
+:: conda-build breaks us. Manual fix:
+set "MSYS2_ARG_CONV_EXCL=/AI;/AL;/OUT;/out"
+:: Delegate to the Unixy script. We need to translate the key path variables
+:: to be Unix-y rather than Windows-y, though.
+set "saved_recipe_dir=%RECIPE_DIR%"
+FOR /F "delims=" %%i IN ('cygpath.exe -u -p "%PATH%"') DO set "PATH_OVERRIDE=%%i"
+FOR /F "delims=" %%i IN ('cygpath.exe -m "%BUILD_PREFIX%"') DO set "BUILD_PREFIX_M=%%i"
```
Comment (JeanChristopheMorinPerso) [inline recipe/bld.bat:8]: I'm guessing `M` stands for msys2? If so, it wouldn't hurt to have it fully written out.

Comment (ifitchet) [reply recipe/bld.bat:8]: Where I've seen such an `_M` involving `cygpath`, it's the "mode": `_U` unix (`/cygdrive/c/users`), `_M` mixed (`C:/Users`) and `_W` windows (`C:\Users`).

File: recipe/build.sh (hunk 2 above)
Comment (ifitchet) [reply recipe/build.sh]: especially as that's setting `IFS` back to its default value.

**Comment by skupr-anaconda** [comment]
Superseded by https://github.com/AnacondaRecipes/gmp-feedstock/pull/4

**Pull request closed without merging**